### ISSUE TITLE
rule no-magic-numbers should support es6 default parameters

### ISSUE THREAD

**Issue by atte-kemppila** [eslint/eslint#5672]
**What version of ESLint are you using?**
v2.4.0

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**What did you do? Please include the actual source code causing the issue.**
```
/* eslint-env es6 */
/* eslint no-magic-numbers: [2, {ignore: [123]}] */

function foo(value = 123) {}  // no errors
function bar(value = 456) {}  // no-magic-numbers error
```

**What actually happened? Please include the actual, raw output from ESLint.**
```
  5:22  error  No magic number: 456  no-magic-numbers

✖ 1 problem (1 error, 0 warnings)
```

**What did you expect to happen?**

In my eyes, default parameters values aren't magical numbers since parameter names already give them context.

And while it's possible to do write something like the code below, I don't see much difference in their magical properties.
```
var magicValue = 789;
function baz(value = magicValue) {}
```

So, I propose that no-magic-number rule to be changed so that it ignores default parameters by default. 

(Or if you don't want change how the rule behaves by default, a new option "ignoreDefaultParams" etc. to be added.)

**Comment by platinumazure** [comment]
Seems logical to me- magic numbers by definition means we need a variable to represent the value, and default assignment does do that with parameters, more or less.

**Comment by ilyavolodin** [comment]
Sounds reasonable to me. What do others think?

**Comment by alberto** [comment]
Agree.

While looking into add this, I came with this cases.

```
var [a, b] = [1, 2];

var [a=1, b=2] = [3];

var {a=1, b=2} = {a: 3};
```

Should we also include them?

**Comment by michaelficarra** [comment]
I very strongly disagree with this. The name given in the parameter list is the meaning of the number within a particular context. The name given to a constant describes the number regardless of any context.

Contrast

```js
function startServer(portNumber = 22) { /* ... */ }
```

with

```js
const SSH_PORT_NUMBER = 22;
function startServer(portNumber = SSH_PORT_NUMBER) { /* ... */ }
```

Notice that 22 is `portNumber` in the context of `startServer`. But that doesn't tell us anything about what 22 means outside that context. The second example makes that clear.

**Comment by alberto** [comment]
Mmmm, good point.

**Comment by alberto** [comment]
@ilyavolodin @platinumazure what's your opinion on @michaelficarra comment?

**Comment by platinumazure** [comment]
Given what @michaelficarra has posted here, I guess I would at the very least want this change to be behind an option (off by default). That way the user gets a choice. I can no longer get behind unconditionally removing the no-magic-numbers check from default parameters.

**Comment by mdebruijne** [comment]
If it is possible add an option to treat any form of constant/variable assignment as a non magic number then that would be great. Whether assignment is based on direct assignment, default parameters, destructuring, logical OR assignment, etcetera.

``` javascript
const portNumber = process.env.NODEJS_PORT_NUMBER || 8080
// 8080 is not a magic number, only assigned to a constant, not being used in code
```

While I can completely understand @michaelficarra preference. To be honest, it is not mine. I always prefer self-containment by assigning constants/variables in the smallest scope possible. I don't necessarily need to know the meaning of a constant/variable outside the scope/context/unit where it is being used.

**Comment by kaicataldo** [comment]
There hasn't been any movement on this issue for a while - time to close this?

**Comment by kaicataldo** [comment]
Whoops, didn't mean to close it yet!

**Comment by alberto** [comment]
Thanks for your interest in improving eslint. Unfortunately, it looks like this issue didn't get enough support from the team and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after 21 days tend to never do it, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.
